server/src/main/java/org/elasticsearch/health/node/DiskHealthIndicatorService.java

```diff
@@ -0,0 +1,469 @@
+            .indices()
+            .entrySet()
+            .stream()
+            .filter(entry -> entry.getValue().contains(IndexMetadata.INDEX_READ_ONLY_ALLOW_DELETE_BLOCK))
+            .map(Map.Entry::getKey)
+            .collect(Collectors.toSet());
+        boolean hasAtLeastOneIndexReadOnlyAllowDeleteBlock = indicesWithBlock.isEmpty() == false;
+        HealthIndicatorDetails details = getDetails(explain, diskHealthInfoMap, clusterState);
+        final HealthStatus healthStatusFromNodes = HealthStatus.merge(
+            diskHealthInfoMap.values().stream().map(DiskHealthInfo::healthStatus)
+        );
+        final HealthStatus healthStatus = hasAtLeastOneIndexReadOnlyAllowDeleteBlock ? HealthStatus.RED : healthStatusFromNodes;
+
+        final HealthIndicatorResult healthIndicatorResult;
+        if (HealthStatus.GREEN.equals(healthStatus)) {
+            healthIndicatorResult = createIndicator(
+                healthStatus,
+                "The cluster has enough available disk space.",
+                details,
+                List.of(),
+                List.of()
+            );
+        } else {
+            /*
+             * Something is unhealthy. We only report on the highest-priority problem. Basically:
+             *
+             * RED > YELLOW
+             * data nodes (including index locks from the cluster state) > master nodes > nodes of all other roles
+             *
+             * The highest priority problem is when we detect that there is an index with a block in the cluster state or a node that has
+             * a data role reporting a RED status (meaning it has exceeded the flood stage). This means that there is likely an index that
+             * cannot be written to.
+             * The next highest priority problem is when we detect that there is a non-data node that has a master role reporting a RED
+             * status (meaning it has exceeded the flood stage). This could lead to problems with cluster formation, discovery, etc.
+             * The next highest priority problem is when we detect that there is a non-data non-master node reporting a RED status
+             * (meaning it has exceeded the flood stage). This could lead to problems with certain features.
+             *
+             * The next tier down is when there are no nodes reporting RED, but some are reporting YELLOW. There will be no index
+             * blocks in the cluster state if we get here because the status would have automatically been RED. Similar to above, we only
+             * report problems with data nodes if there are any data nodes reporting YELLOW (the high watermark has been exceeded). If
+             * there are no data nodes reporting YELLOW, we report problems with any master nodes reporting YELLOW. If there are no
+             * master nodes reporting YELLOW, we report on the non-data non-master nodes that report YELLOW.
+             */
+            if (hasAtLeastOneIndexReadOnlyAllowDeleteBlock) {
```

Comment:
Can we have a more manageable name here? 
maybe `isBlockedIndex` / `clusterHasBlockedIndex` or something like that?